homeassistant/components/repetier/__init__.py

```diff
@@ -0,0 +1,262 @@
+}
+
+SENSOR_SCHEMA = vol.Schema({
+    vol.Optional(CONF_MONITORED_CONDITIONS, default=list(SENSOR_TYPES)):
+        vol.All(cv.ensure_list, [vol.In(SENSOR_TYPES)]),
+    vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+})
+
+CONFIG_SCHEMA = vol.Schema({
+    DOMAIN: vol.All(cv.ensure_list, [vol.Schema({
+        vol.Required(CONF_API_KEY): cv.string,
+        vol.Required(CONF_HOST): cv.string,
+        vol.Optional(CONF_PORT, default=3344): cv.port,
+        vol.Optional(CONF_NAME, default=DEFAULT_NAME): cv.string,
+        vol.Optional(CONF_SENSORS, default={}): SENSOR_SCHEMA,
+    })], has_all_unique_names),
+}, extra=vol.ALLOW_EXTRA)
+
+
+def setup(hass, config):
+    """Set up the Repetier Server component."""
+    import pyrepetier
+
+    hass.data[REPETIER_API] = {}
+
+    for repetier in config[DOMAIN]:
+        _LOGGER.debug("Repetier server config %s", repetier[CONF_HOST])
+
+        url = "http://{}".format(repetier[CONF_HOST])
+        port = repetier[CONF_PORT]
+        api_key = repetier[CONF_API_KEY]
+
+        client = pyrepetier.Repetier(
+            url=url,
+            port=port,
+            apikey=api_key)
+
+        printers = client.getprinters()
+
+        if not printers:
+            return False
+
+        sensors = repetier[CONF_SENSORS][CONF_MONITORED_CONDITIONS]
+        api = PrinterAPI(hass, client, printers, sensors,
+                         repetier[CONF_NAME], config)
+        api.update()
+        track_time_interval(hass, api.update, SCAN_INTERVAL)
+
+        hass.data[REPETIER_API][repetier[CONF_NAME]] = api
+
+    return True
+
+
+class PrinterAPI:
+    """Handle the printer API."""
+
+    def __init__(self, hass, client, printers, sensors, conf_name, config):
+        """Set up instance."""
+        self._hass = hass
+        self._client = client
+        self.printers = printers
+        self.sensors = sensors
+        self.conf_name = conf_name
+        self.config = config
+        self._known_entities = []
+
+    def get_data(self, printer_id, sensor_type, temp_id):
+        """Get data from the state cache."""
+        printer = self.printers[printer_id]
+        methods = API_PRINTER_METHODS[sensor_type]
+        for prop, offline in methods['offline'].items():
+            state = getattr(printer, prop)
+            if state == offline:
+                # if state matches offline, sensor is offline
+                return None
+
+        data = {}
+        for prop, attr in methods['state'].items():
+            prop_data = getattr(printer, prop)
+            if attr == 'temp_data':
+                temp_methods = methods['temp_data']
+                for temp_prop, temp_attr in temp_methods.items():
+                    data[temp_attr] = getattr(prop_data[temp_id], temp_prop)
+            else:
+                data[attr] = prop_data
+        return data
+
+    def update(self, now=None):
+        """Update the state cache from the printer API."""
+        self._load_entities()
```

Comment:
Do we also need to call `_client.getprinters` to get any new printers?

Reply:
I think we need to move this below the iteration of printers and `printer.get_data()`. Otherwise the printer data won't have updated.

Reply:
Done - seemed to work without moving the line, but to eliminate doubt I moved it below get_data.